File: 930-binary-subarrays-with-sum/930-binary-subarrays-with-sum.cpp
class Solution {
public:
    int numSubarraysWithSum(vector<int>& nums, int k) {
        int sum=0;
        map<int,int> mp;
        mp[0]++;
        int res=0;
        for(int i=0;i<nums.size();i++)
        {
            sum+=nums[i];
            if(mp.find(sum-k)!=mp.end())
            {
                res += mp[sum-k];
            }
            mp[sum]++;
        }
        return res;
    }
};

File: 930-binary-subarrays-with-sum/930-binary-subarrays-with-sum.java
class Solution {
    public int numSubarraysWithSum(int[] nums, int goal) {
        int l=0,r=0;
        int n = nums.length;
        int cursum =0 , res=0;
        while(l<n && r<n)
        {
            cursum+= nums[r];
            if(cursum<goal)
                r++;
            else if(cursum>goal)
            {
                l++;
                r = l;
                cursum=0;
            }
            else 
            {
                res++;
                r++;
                if(r==n){
                    l++;
                    r = l;
                    cursum=0;
                }
            }
        }
        return res;
    }
}